nixos/modules/services/networking/smokeping.nix

```diff
@@ -311,15 +328,23 @@ in
         ${cfg.package}/bin/smokeping --check --config=${configPath}
         ${cfg.package}/bin/smokeping --static --config=${configPath}
       '';
-      script = "${cfg.package}/bin/smokeping --config=${configPath} --nodaemon";
     };
     systemd.services.thttpd = mkIf cfg.webService {
-      wantedBy = [ "multi-user.target"];
+      requiredBy = [ "multi-user.target"];
       requires = [ "smokeping.service"];
-      partOf = [ "smokeping.service"];
       path = with pkgs; [ bash rrdtool smokeping thttpd ];
-      script = ''thttpd -u ${cfg.user} -c "**.fcgi" -d ${smokepingHome} -p ${builtins.toString cfg.port} -D -nos'';
-      serviceConfig.Restart = "always";
+      serviceConfig = {
+        Restart = "always";
+        ExecStart = lib.concatStringsSep " " (lib.concatLists [
```

Comment:
What is the necessity of `concatLists` on a list of single entry lists? Why not use just a list?

Reply:
It allows to do things like I did below,

```
(lib.optional (cfg.host != null) "-h ${cfg.host}")
```

which itself is a list (either `[]` or one with contents) without having to use `++`, which is not `diff` / merge conflict avoidance friendly:

In the below you can delete any line (including the first or last) and have it be a single-line diff.

The closest to it without `concat` is

```
      (
        [
          "${pkgs.thttpd}/bin/thttpd"
          "-u ${cfg.user}"
          ''-c "**.fcgi"''
          "-d ${smokepingHome}"
        ]
        ++ (lib.optional (cfg.host != null) "-h ${cfg.host}")
        ++ [
          "-p ${builtins.toString cfg.port}"
          "-D -nos"
        ]
      )
```

which doesn't allow deleting the first `[]` block without also having to touch the `++` on the next line, because `++` is a binary operator and list elements aren't. `++`-based syntax tends to introduce merge conflicts in nixpkgs.

So it's a tradeoff between slight structural wrapping, and slightly improved merge conflict avoidance.

I don't' have a big preference either way.

Reply:
Makes sense to me, thanks for the thorough explanation!